### ISSUE TITLE
How to check if #external predicates are defined properly?

### ISSUE THREAD

**Issue by ejgroene** [potassco/clingo#503]
My code in `lib.lp` uses rules with externally defined inputs. How can I check if these inputs are properly defined?

I have a `client.lp` that defines `inputs` and later provides values for these input at different steps:

    #include "lib.lp".

    def_input(switch_1,   on, "S1E").
    def_input(switch_1,  off, "S1A").

    input(0, "S1E").  % T=0
    input(1, "S1A").  % T=1

In `lib.lp` I use the inputs on a functional level, referring to them as `on` and `off`:

    time(0..2).
    #defined def_input/3.
    #external input(T, Name)  :  time(T),  def_input(Device, Function, Name).

    switched_on(T)   :-  def_input(Device,  on, Name),  input(T, Name).
    switched_off(T)  :-  def_input(Device, off, Name),  input(T, Name).

Now, if the client did not specify `on` and `off` properly, the logic does not work.

How can I make sure that `def_input(switch_1,   on, "S1E")` is defined for `on` and similarly for `off`?

**Comment by ejgroene** [comment]
Thanks for your answer, today I am (re) trying mentioned approach.

**Comment by rkaminsk** [comment]
You could also nave a look at the following example:

https://github.com/potassco/clingo/tree/master/examples/clingo/cannot

**Comment by ejgroene** [comment]
Thank you. I have tried to use this, but unfortunately it suffers the same vulnerability as my previous attempts. Let me explain:

If/when gringo discovers a literal that can't be true, it will not propagate the theory atom, just as it will not add the rule or call Python functions. I have tried now:

Using @all to register the test case and predicate `assert` which **absence** signals problems:

    assert(@all("a test case))  :-  device(D),  def_input(D,  on, _).


Use predicate `problem` which **presence** signals problems:

    problem( "no 'on' for", D)  :-  device(D), not def_input(D,  on, _).


Use a theory atom for the head of a constraint:

    &cannot {"no 'on' for", D}   :-  device(D), not def_output(D, stuur, _).


The returning problem is that gringo will optimize away the whole rule in certain cases, so it does work, as long as I am very careful. For example, the negation above can sometimes be written like:

    problem( "no 'on' for", D)  :-  device(D), { def_input(D,  on, _) } = 1.

It would be nice if `gringo` would not optimize away certain things in a test mode, or something like that.

That being said, I still love this way of writing asserts and thus unit testing my programs. I couldn't live without it.
